vim-patch:8.1.2304,8.1.2309,8.1.2319,8.1.2321

dstein64 opened the pull request:
This PR ports Vim's `getmousepos()` function to Neovim.

The patches were not applied as-is. The changes include:

* Update code formatting.
* Move `f_getmousepos` from `mouse.c` to `funcs.c`.
* Remove handling for popup windows.
* Add handling for floating windows with borders. The code adds handling to consider borders as part of the window for the purpose of computing `winrow` and `wincol`.
* Add tests.

~Commit df48a6ffdebfca5b5e682119264eba0ac9b1dbde updates `getmousepos()` to consider unfocusable floating windows. This matches how `getmousepos()` works with popup windows in Vim, as popup windows are unfocusable. The update adds an `include_unfocusable` argument to `mouse_find_win` for specifying whether that function should consider unfocusable floating windows. The value `false` was passed in all cases outside `f_getmousepos` so that existing behavior is maintained.~


#### vim-patch:8.1.2304: cannot get the mouse position when getting a mouse click

Problem:    Cannot get the mouse position when getting a mouse click.
Solution:   Add getmousepos().
https://github.com/vim/vim/commit/db3a205147ce2c335d5c2181c1f789277f8775b0


#### vim-patch:8.1.2309: compiler warning for argument type

Problem:    Compiler warning for argument type.
Solution:   Use linenr_T and cast to varnumber_T. (John Marriott)
https://github.com/vim/vim/commit/abe12a1a4fce36bfa5dea3a0ce85603432d1905b


#### vim-patch:8.1.2319: compiler warning for int size

Problem:    Compiler warning for int size.
Solution:   Add typecast. (Mike Williams)
https://github.com/vim/vim/commit/07a63d86338476bafbd1a3ec462672df92666498


#### vim-patch:8.1.2321: cannot select all text with the mouse 

Problem:    Cannot select all text with the mouse. (John Marriott)
Solution:   Move limiting the mouse column to f_getmousepos().
            (closes https://github.com/vim/vim/issues/5242)
https://github.com/vim/vim/commit/0a5aa7b28a39507260acb15c1ef698a33c855cc1

dstein64 committed:
vim-patch:8.1.2304: cannot get the mouse position when getting a mouse click

Problem:    Cannot get the mouse position when getting a mouse click.
Solution:   Add getmousepos().
https://github.com/vim/vim/commit/db3a205147ce2c335d5c2181c1f789277f8775b0

dstein64 committed:
test: add a test for getmousepos()

dstein64 commented on runtime/doc/eval.txt:
```diff
@@ -4704,6 +4706,36 @@ getmatches([{win}])					*getmatches()*
+		are zero.
+
+		When the position is after the text then "column" is the
+		length of the text in bytes.
```
This documentation was ported directly from Vim. The documentation currently does not match the behavior, since `getmousepos().column` in this case would be "the length of the text in bytes" _plus one_. Vim's help will be [updated](https://github.com/vim/vim/issues/8474#issuecomment-869953539) to correct this. I haven't made any updates here, with the intent that the future update to the Vim documentation be ported to Neovim, with the text kept in sync.

dstein64 replied:
The documentation was corrected in Vim as part of https://github.com/vim/vim/commit/90df4b9d423485f7db16e3a65cab4f14edc815ae. I've applied this same update in 3dbf80e250bde2edc8587a8c75073b062547c415.

bfredl commented:
> Commit df48a6f updates getmousepos() to consider unfocusable floating windows. This matches how getmousepos() works with popup windows in Vim, as popup windows are unfocusable.

I don't think this is correct. The point of the "focusable" option is to expliticly **control** whether the window receives focus or not, while vim lets plugins use various ad-hockeries to emulate some forms of focus of popups, like filter callbacks. So I think this should respect our own definition of "focusable".

dstein64 commented:
Thanks for the feedback @bfredl!

df48a6f was added so that if a user clicks on a non-focusable window, `getmouepos()` would return data corresponding to that non-focusable window (its window ID, the clicked window row and column, and the clicked text line and column). Without that, if a user clicks a non-focusable window, `getmousepos()` would return data corresponding to the focusable window underneath the non-focusable window. Besides the `getmousepos()` handling, all other aspects of window focus were kept the same with that commit.

This was added so that the behavior would match Vim, where the data returned by `getmousepos()` would correspond to clicks on non-focusable popups, as opposed to the window underneath.

> "So I think this should respect our own definition of "focusable"."

@bfredl, in case those details provide information that may not have been clear in my original comment, which of the following do you think is preferable?

1. `getmousepos()` considers clicks on non-focusable windows (the current implementation), as opposed to treating the click on the window underneath, matching Vim's behavior.
2. `getmousepos()` ignores clicks on non-focusable windows, considering the window underneath (revert df48a6f), differing from Vim's behavior.

Alternatively, we could have an optional argument to `getmousepos()` to toggle this behavior (this would seemingly be more applicable outside this PR, whose main purpose is to port the patch from Vim). In that case, I think it would be preferable to have the default match Vim, so that plugin authors can avoid separate code pathways for Vim and Neovim.

bfredl commented:
> This was added so that the behavior would match Vim, where the data returned by getmousepos() would correspond to clicks on non-focusable popups, as opposed to the window underneath. ...  In that case, I think it would be preferable to have the default match Vim, so that plugin authors can avoid separate code pathways for Vim and Neovim (

But the default behavior of floats, i e focusable=true, already matches the behavior with vim popups w.r.t mouse clicks. `focusable` option does not exist in vim, so here is nothing here to match. A plugin setting `focusable=false` does so because they **want** the window to be excluded from focus and all its implications. If this is not intended, this option should not be set. It doesn't make sense to prevent plugins to opt-in to a behavior just because it isn't possible in vim.

dstein64 commented:
> "But the default behavior of floats, i e focusable=true, already matches the behavior with vim popups w.r.t mouse clicks."

In Vim, for example, if I create a popup with `call popup_create('hello', {})`, then click on that popup, focus will not go to that popup. This behavior more closely matches Neovim's `focusable=false`. However, a difference between `focusable=false` in Neovim and Vim popups is that clicking on a Vim popup ignores the click (i.e., no cursor movement to a new place in the window underneath), whereas Neovim treats the click as having occurred on the window underneath and possibly moves the cursor to a new spot in that window.

bfredl commented:
I didn't say the behavior was identical in every aspect, this just about the focusing part.

>  However, a difference between focusable=false in Neovim and Vim popups is that clicking on a Vim popup ignores the click (i.e., no cursor movement to a new place in the window underneath), 

So the click is still _received_ by the popup rather than the window beneath, which matches the neovim behavior of `focusable=true`. Now, vim will not enter normal mode inside the popup, as the powers that be have decided this is not desired, but the mouse click in the popup can still be _used_ for other purposes, such as builtin drag, resize and close behaviors, or by plugin using some filter callback thing probably.

Regardless what we do the behavior will not match vim precisely. In the end we have to have confidence in our own design choices, and provide a solid foundation for plugins (and possibly, compatibility layers). And I think the API is easier to reason about and for plugins to use if `focusable=false` is consistently interpreted as disregarding the float (at least by default).

dstein64 commented:
> "And I think the API is easier to reason about and for plugins to use if focusable=false is consistently interpreted as disregarding the float (at least by default)."

👍 I'll revert that commit.

dstein64 committed:
vim-patch:8.1.2309: compiler warning for argument type

Problem: Compiler warning for argument type.
Solution: Use linenr_T and cast to varnumber_T. (John Marriott)
https://github.com/vim/vim/commit/abe12a1a4fce36bfa5dea3a0ce85603432d1905b

dstein64 committed:
vim-patch:8.1.2319: compiler warning for int size

Problem: Compiler warning for int size.
Solution: Add typecast. (Mike Williams)
https://github.com/vim/vim/commit/07a63d86338476bafbd1a3ec462672df92666498

dstein64 committed:
vim-patch:8.1.2321: cannot select all text with the mouse

Problem: Cannot select all text with the mouse. (John Marriott)
Solution: Move limiting the mouse column to f_getmousepos().
          (closes https://github.com/vim/vim/issues/5242)
https://github.com/vim/vim/commit/0a5aa7b28a39507260acb15c1ef698a33c855cc1

dstein64 commented:
PR #14944 addresses the issue causing Windows test/build failures in this PR.

dstein64 committed:
chore: remove empty line

This update was part of vim-patch:4c29502.
https://github.com/vim/vim/commit/4c295027a426986566cd7a76c47a6d3a529727e7
The other parts of that patch were applied to Neovim in ad6bb38.

dstein64 committed:
docs: add 'plus one' to correct getmousepos docs

This change was applied to Vim as part of 90df4b9.
https://github.com/vim/vim/commit/90df4b9d423485f7db16e3a65cab4f14edc815ae

janlazo commented:
Wrong commit format so this cannot be merged.

dstein64 commented:
> "Wrong commit format for vim-patch so this cannot be merged."

I've updated the patch commit format to follow the convention. Thanks!

Linting is currently failing because there are three commits in this PR that don't have `vim-patch` in their commit message, a requirement added in #15129. The earlier `vim-patch` commits closely match the corresponding Vim commits, and they're followed by a few related supplemental updates that were made in addition to the patches themselves.

I can 1) squash the supplemental updates into the corresponding patch commits, 2) add `vim-patch` to the commit message for the supplemental updates to satisfy the linter, 3) relax the linting requirement, or 4) use some other workaround. **Please let me know the preferred way to proceed.**

clason commented:
> Linting is currently failing because there are three commits in this PR that don't have the vim-patch type in their commit message, a requirement added in #15129. The four vim-patch commits closely match the corresponding Vim commits, and they're followed by a few related supplemental updates that were made in addition to the patches themselves.

Yes, that's a known issue with the commitlinter: you currently can't mix `vim-patch` and non-`vim-patch` commits in one PR (since they are linted by different scripts). Just ignore the lint failure.

janlazo commented on src/nvim/testdir/test_functions.vim:
```diff
@@ -1576,4 +1576,122 @@ func Test_default_arg_value()
   call assert_equal('msg', HasDefault())
 endfunc
 
+func Test_getmousepos()
```
This is not from Vim so this will introduce merge conflicts.

Port this to a lua functional test or make it obvious via comments that it is for Neovim only.

dstein64 replied:
I've ported the test to Lua (ae97f25e9d10440cb390ce8b9f9cc50adcc71e02). Thanks!
